docs/source/extending/overloading-guide.rst

```diff
@@ -0,0 +1,251 @@
+    5.88 ms ± 43.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
+
+And since you use this function, very often in your ``jit`` decorated
+functions, for example in ``myalgorithm``, you choose to use to ``overload`` in
+an attempt to accelerate things.
+
+.. code:: python
+
+    @njit
+    def myalgorithm(a, x):
+        # algorithm code
+        ham.set_to_x(a, x)
+        # algorithm code
+
+Inspired by the template above, your implementation might look something like:
+
+.. literalinclude:: myjitmodule.py
+
+
+
+And when you go to profile it, you find yourself pleasantly surprised:
+
+.. code:: pycon
+
+    In [1]: import numpy as np
+
+    In [2]: import myjitmodule
+
+    In [3]: a = np.arange(100000)
+
+    In [4]: %timeit myjitmodule.myalgorithm(a, 1)
+    17.6 µs ± 327 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
+
+But of course, your implementation doesn't generalize to your colleagues
+use-case, who would like to use ``set_to_x`` with a floating point number
+instead:
+
+.. code:: pycon
+
+    In [4]: myjitmodule.myalgorithm(a, 1.0)
+    TypingError: Failed in nopython mode pipeline (step: nopython frontend)
+    Invalid use of Function(<function set_to_x at 0x11aea71e0>) with argument(s) of type(s): (array(int64, 1d, C), float64)
+    * parameterized
+    In definition 0:
+        All templates rejected with literals.
+    In definition 1:
+        All templates rejected without literals.
+    This error is usually caused by passing an argument of a type that is unsupported by the named function.
+    [1] During: resolving callee type: Function(<function set_to_x at 0x11aea71e0>)
+    [2] During: typing of call at /Users/vhaenel/git/numba/docs/source/extending/myjitmodule.py (25)
+
+
+    File myjitmodule.py", line 25:
+    def myalgorithm(a, x):
+        # algorithm code
+        mymodule.set_to_x(a, x)
+        # algorithm code
+
+
+Providing multiple implementations and dispatching based on types
+=================================================================
+
+As you saw above, the overload implementation for  ``set_to_x`` function
+doesn't accept floating point arguments. Let's extended the specification of
+the function as follows:
+
+* The numerical type of the array ``arr`` must match the numerical type of the
+  scalar ``x`` argument, i.e. if ``arr`` is of type ``int64``, then ``x`` must
+  be of this type too.
+* Only integer and floating-point types are to be supported
+* For the floating-point type, no ``nan`` values are allowed in ``arr`` and if
+  such a value is encountered, an appropriate ``ValueError`` should be raised.
+* If a tuple is used instead of an array, a custom error message with a hint
+  for the user should be issued.
+
+The resulting implementation could look as follows:
+
+.. literalinclude:: myjitmodule2.py
+
+As you can see, the typing checking code has been increased significantly to
+match the new requirements. Also, multiple implementations---one for integers
```

Comment:
The last line in this snippet might look surprising to folk who are starting out.

```python
    if isinstance(arr, types.Array):
        # validate that arr and x have the same type
        if arr.dtype == x:
```

Reply:
Do you think it would suffice to expand the comment a little along the lines of:

```
(The arr argument is an array and the type of the vlaues is stored as the 'dtype' attribute. The x argument on the other hand is a scalar and thus x is already the type here.)
```

The key is, there are quasi two parts to such implementations: the type inference stage and the actual implementation stage. And, the two stages are hidden in decorator/closure implementation. I.e. you do the type checking in the function itself and then return another function with the concrete implementation. The (Python) types of the arguments are different in both cases, in the first they are the Numba types and in the second the real arguments. I am not yet sure how to drive this point home. Maybe take a step back and explain or link to the concept of decorators and closures? What do you think?

Reply:
Yeah, I think the expanded comment is perfect here.  Personally, I would keep it at that.

Reply:
Thanks!